### PR TITLE
[AIRFLOW-5691] Rewrite Dataproc operators to use python library

### PR DESCRIPTION
**Pull request by turbaszek**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5691

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
This PR introduces refactored Dataproc operators that use python client for Dataproc instead of discovery API. Job operators like `DataProcHadoopOperator` are going to be deprecated and user should use generic `DataprocSubmitJobOperator`. In general new operators will accept only request objects like `Cluster` or `Job`. I've added helper methods that will help users to generate those objects using their old operators.

This PR also decouples hook, operators any additional Dataproc helpers.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/gcp/example_dags/example_dataproc.py [hunk 1]
```diff
@@ -41,69 +40,130 @@
...
     default_args={"start_date": airflow.utils.dates.days_ago(1)},
     schedule_interval=None,
 ) as dag:
     create_cluster = DataprocClusterCreateOperator(
-        task_id="create_cluster",
-        cluster_name=CLUSTER_NAME,
-        project_id=PROJECT_ID,
-        num_workers=2,
-        region=REGION,
+        task_id="create_cluster", project_id=PROJECT_ID, cluster=CLUSTER, region=REGION
```
Comment (mik-laj) [inline airflow/gcp/example_dags/example_dataproc.py]: Have you tested this change with the original DAG?

File: airflow/gcp/hooks/dataproc.py [hunk 2]
```diff
@@ -498,215 +836,59 @@ def submit(
...
-        return _DataProcJobBuilder(
-            self.project_id,
-            task_id,
-            cluster_name,
-            job_type,
-            properties
+        # TODO: Remover one day
+        warnings.warn(
+            "This method is deprecated. Please use `submit_job`",
+            DeprecationWarning,
```
Comment (mik-laj) [inline airflow/gcp/hooks/dataproc.py]:
```suggestion
            DeprecationWarning,
            stacklevel=2
```

File: airflow/gcp/example_dags/example_dataproc.py (hunk 1 above)
Comment (turbaszek) [reply airflow/gcp/example_dags/example_dataproc.py]: Yes, that was how I checked backwards compatibility. Channing example DAG was last thing I did. If you wish I can change the DAG in separate PR.

**Comment by turbaszek** [comment]
@mik-laj, @potiuk can you take a look at the Travis? I am not sure if this is something flaky or not. 
https://api.travis-ci.org/v3/job/599993904/log.txt

**Comment by mik-laj** [comment]
It looks like some kind of dependencies has exploded again.

**Comment by mik-laj** [comment]
I restarted failed job.

**Review by potiuk — CHANGES_REQUESTED** [review]
Just one change -> providers/google/cloud

**Review by potiuk — APPROVED** [review]

**Comment by dossett** [comment]
This seems to have made some substantive changes to the behavior of the operators, was that intended?  For example, functionality to detect and reattach to a running job is gone.  See the code starting here for functionality that wasn't ported. https://github.com/apache/airflow/commit/d633d3ac44c395e6c43cd388f98fba1ce1c435a3#diff-0be5a6bccaef6a25d9ff5d63a92a12f0L64-L65

@potiuk @mik-laj @nuclearpinguin

**Comment by dossett** [comment]
That link may not expand well.  It's to line 64 in the `airflow/gcp/hooks/dataproc.py` that was renamed as part of this change.

**Comment by dossett** [comment]
The functionality that was apparently lost is AIRFLOW-3211

**Comment by ghost** [comment]
@dossett, the functionality added in AIRFLOW-3211 actually broke the behavior of the dataproc hook and made a few 1.10.x releases unusable for dataproc users. The problem is that the hook only uses the task ID part of the dataproc job ID when looking for previous invocations of the job, so if dataproc history still has jobs corresponding to any of the previous dag runs, the dataproc hook doesn't execute the job. A proper way to implement this would be to associate dataproc jobs with particular dag runs by e.g. embedding a dag run id hash in the dataproc job id. In any case the functionality added in  AIRFLOW-3211 has to be optional. In our experience, users expect dataproc jobs to be re-executed when they re-execute the task, and this new behavior creates a lot of confusion.

**Comment by dossett** [comment]
@digger I take your points about the original change. Do you know if reverting that functionality was an intentional part of this PR?

**Comment by digger** [comment]
@dossett, I don't know if that was intentional. I just shared my feedback on changes made in AIRFLOW-3211. In the company I work for we use Airflow and we had to patch Airflow 1.10.7 and 1.10.9, reverting changes from AIRFLOW-3211, in order to make the dataproc functionality work.

**Comment by nuclearpinguin** [comment]
@digger is right, Dataproc was not working (tested using example DAG). I've tried to do my best to preserve as much backward compatibility as I could. The present implementation is much clearer and uses Google's python library. Without operators - hooks coupling it should be easier to add any needed changes.

I am happy to review PR that adds missing functionality @dossett :)

**Comment by dossett** [comment]
@nuclearpinguin It looks like functionality added by AIRFLOW-3149 (to more gracefully handle the creation of clusters in an ERROR state) was also reverted, was that also intentional?  Was there a JIRA or mailing list discussion of the functional changes that accompanied this API migration? (I'm seriously asking, I searched but did not find anything.)

**Comment by nuclearpinguin** [comment]
There's a JIRA: https://issues.apache.org/jira/browse/AIRFLOW-5691

No change was intentional.  Also, this change was not backported to 1.10.x .

So, from what I understand the submit job operator should:
- check if job with provided id already exists
- if yes then the existing job should be used
- if no create new job

Doesn't `request_id` work like that?
> If the server receives two SubmitJobRequest requests with the same id, then the second request will be ignored and the first Job created and stored in the backend is returned.

https://googleapis.dev/python/dataproc/latest/gapic/v1/api.html#google.cloud.dataproc_v1.JobControllerClient.submit_job

@dossett @digger

**Comment by dossett** [comment]
@turbaszek I hope to return to this topic once world events stabilize and my kids return to school.  Cheers! -Aaron

**Pull request merged**